snuba/clusters/cluster.py

```diff
@@ -0,0 +1,104 @@
+from abc import ABC, abstractmethod
+from typing import Set
+
+from snuba import settings
+from snuba.clickhouse.native import ClickhousePool, NativeDriverReader
+from snuba.clickhouse.query import ClickhouseQuery
+from snuba.clusters.storage_sets import StorageSetKey
+from snuba.reader import Reader
+
+
+class Cluster(ABC):
+    """
+    A cluster is responsible for managing a collection of database nodes.
+
+    Clusters are configurable, and will be instantiated based on user defined settings.
+
+    Each storage must be mapped to a cluster via a storage set, which defines
+    the storages that must be located on the same cluster.
+
+    In future, clusters will also be responsible for co-ordinating commands that
+    need to be run on multiple hosts that are colocated within the same cluster.
+    The cluster will expose methods for:
+        - bootstrap
+        - migrate
+        - cleanup
+        - optimize
+    """
+
+    def __init__(self, storage_sets: Set[str]):
+        self.__storage_sets = storage_sets
+
+    def get_storage_sets(self) -> Set[StorageSetKey]:
+        return {StorageSetKey(storage_set) for storage_set in self.__storage_sets}
+
+    @abstractmethod
+    def get_reader(self) -> Reader[ClickhouseQuery]:
+        raise NotImplementedError
+
+
+class ClickhouseCluster(Cluster):
+    """
+    ClickhouseCluster provides a reader and Clickhouse connections that are shared by all
+    storages located on the cluster
+    """
+
+    def __init__(self, host: str, port: int, http_port: int, storage_sets: Set[str]):
+        self.__host = host
+        self.__port = port
+        self.__clickhouse_rw = ClickhousePool(host, port)
+        self.__clickhouse_ro = ClickhousePool(
+            host, port, client_settings={"readonly": True},
+        )
+        self.__reader: Reader[ClickhouseQuery] = NativeDriverReader(
+            self.__clickhouse_ro
+        )
+        super().__init__(storage_sets)
```

Comment:
nit: putting the parent class initialization at the beginning makes it teasier to see the correctness of the method.